sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/providers/BigQueryDirectReadSchemaTransformProvider.java

```diff
@@ -0,0 +1,206 @@
+
+    @Nullable
+    public abstract BigQueryServices getBigQueryServices();
+
+    /** Builder for the {@link BigQueryDirectReadSchemaTransformConfiguration}. */
+    @AutoValue.Builder
+    public abstract static class Builder {
+      public abstract Builder setQuery(String query);
+
+      public abstract Builder setTableSpec(String tableSpec);
+
+      public abstract Builder setRowRestriction(String rowRestriction);
+
+      public abstract Builder setSelectedFields(List<String> selectedFields);
+
+      public abstract Builder setBigQueryServices(BigQueryServices bigQueryServices);
+
+      /** Builds a {@link BigQueryDirectReadSchemaTransformConfiguration} instance. */
+      public abstract BigQueryDirectReadSchemaTransformConfiguration build();
+    }
+  }
+
+  /**
+   * A {@link SchemaTransform} for BigQuery Storage Read API, configured with {@link
+   * BigQueryDirectReadSchemaTransformConfiguration} and instantiated by {@link
+   * BigQueryDirectReadSchemaTransformProvider}.
+   */
+  private static class BigQueryDirectReadSchemaTransform implements SchemaTransform {
+    private final BigQueryDirectReadSchemaTransformConfiguration configuration;
+
+    BigQueryDirectReadSchemaTransform(
+        BigQueryDirectReadSchemaTransformConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public PTransform<PCollectionRowTuple, PCollectionRowTuple> buildTransform() {
+      return new BigQueryDirectReadPCollectionRowTupleTransform(configuration);
+    }
+  }
+
+  static class BigQueryDirectReadPCollectionRowTupleTransform
+      extends PTransform<PCollectionRowTuple, PCollectionRowTuple> {
+    private final BigQueryDirectReadSchemaTransformConfiguration configuration;
+
+    BigQueryDirectReadPCollectionRowTupleTransform(
+        BigQueryDirectReadSchemaTransformConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public PCollectionRowTuple expand(PCollectionRowTuple input) {
+      checkArgument(
+          input.getAll().isEmpty(),
+          String.format(
+              "Input to %s is expected to be empty, but is not.", getClass().getSimpleName()));
+
+      BigQueryIO.TypedRead<TableRow> read = createDirectReadTransform();
+
+      PCollection<TableRow> tableRows = input.getPipeline().apply(read);
+      Schema schema = tableRows.getSchema();
+      PCollection<Row> rows =
+          tableRows.apply(
+              MapElements.into(TypeDescriptor.of(Row.class))
+                  .via((tableRow) -> BigQueryUtils.toBeamRow(schema, tableRow)));
+
+      return PCollectionRowTuple.of(OUTPUT_TAG, rows.setRowSchema(schema));
+    }
+
+    BigQueryIO.TypedRead<TableRow> createDirectReadTransform() {
+      BigQueryIO.TypedRead<TableRow> read =
+          BigQueryIO.readTableRowsWithSchema().withMethod(TypedRead.Method.DIRECT_READ);
+
+      if (!Strings.isNullOrEmpty(configuration.getQuery())) {
```

Comment:
we should validate that at least query or tablespec is populated

Reply:
Validation is already done in BigQueryIO, should there be one here too? 

Bigger question: to what extent should a configuration class validation be similar to that of the actual connector? e.g. should we validate that the tablespec is in the right `<project>:<dataset>.<table>` format, then should we verify the table exists?

Reply:
There is probably some middle ground, maybe validate everything short of making an API call?

Reply:
true. We probably don't need full validation

Reply:
Thanks, PTAL